woocommerce/woocommerce
# SDK to control product feature usage

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:



This PR implements a set of functions for fetching product restrictions from WooCommerce.com to be used by extensions.

- The product restrictions will be defined on WooCommerce.com.
- WooCommerce Core will only try to process minimum required restrictions.
- Any custom product restriction will be a contract between WooCommerce.com and plugin developers.
- The plugin developers needs to call `\WC_Product_Usage::get_rules_for_product( <product_id>:int )` function to fetch product restrictions for a given product.
- WooCommerce Core, will only return product restrictions from above function, only if these restrictions needs to be applied immediately. 



Closes # .



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Checkout this branch.
2. Install Composite Products.
4. Make sure the license for the selected product is not expired and product is connected
5. Goto the wp shell of your local store `pnpm -- wp-env run cli wp shell`
6. Run `\WC_Product_Usage::get_rules_for_product( 216836 );` in wp shell, which should return `NULL`
7. Check following cases using the same method above.

Test Case  | Expected Output 
-- | --
Product has an active connected license | NULL
Product subscription is connected and expired. | `WC_Product_Usage_Rule_Set`
Product subscription is not connected | `WC_Product_Usage_Rule_Set`
No subscription for the product at all (expired or active) | `WC_Product_Usage_Rule_Set`

Example`WC_Product_Usage_Rule_Set` object
```
wp> WC_Product_Usage::get_rules_for_product( 216836 )
=> object(WC_Product_Usage_Rule_Set)#6778 (1) {
  ["rules":protected]=>
  array(2) {
    ["name"]=>
    string(18) "Composite Products"
    ["regular_price"]=>
    string(2) "79"
  }
}
wp>
```



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 
Add methods required by extensions to control product feature usage based on subscription status.
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
